### PR TITLE
KER-320 : Job definition not updated after changed in configuration

### PR DESCRIPTION
**Pull request by aboughzela**
Update persisted job data 
- Updated jobDetail after changed in configuration
- Remove job if the same job key (jobName, JobGroup) is removed from configuration

### CODE REVIEW

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java [hunk 1]
```diff
@@ -645,6 +664,10 @@ public void start()
    {
       try
       {
+         //Remove the list persisted jobs (jobs removed from the conf)
+         for (JobKey jobKey : persistedJobKeyList ){
+            scheduler_.deleteJob(jobKey);
```
Comment (thomasdelhomenie) [inline exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:674]: Do you remove all jobs at each startup here ?

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java [hunk 2]
```diff
@@ -692,4 +715,20 @@ private String getGroupName(String initialGroupName)
          gname = containerName_ + ":" + initialGroupName;
       return gname;
    }
+
+   /**
+    * Return the list of existed jobs key
+    */
+   private void getAllJobKey() {
```
Comment (thomasdelhomenie) [inline exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java]: We should either rename it to "loadAllJobKeys", or to make it return the list of job keys. It is weird to have a get which returns nothing.

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java [hunk 3]
```diff
@@ -692,4 +715,20 @@ private String getGroupName(String initialGroupName)
...
+      try {
+         List<String> jgroups = scheduler_.getJobGroupNames();
```
Comment (thomasdelhomenie) [inline exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:729]: The name is quite confusing :) We should rename it to jobGroupNames instead IMO.

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java (hunk 1 above)
Comment (thomasdelhomenie) [inline exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:674]: You could use deleteJobs(persistedJobKeyList) instead.

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java [hunk 4]
```diff
@@ -163,16 +151,27 @@ private void scheduleJob(JobDetail jobDetail, Trigger trigger) throws SchedulerE
     * return <code>false</code> if and only if the persisted job is the exact same job as the provided one
     * otherwise it will raise a {@link SchedulerException}
     */
-   private boolean registerJob(JobDetail jobDetail) throws SchedulerException
+   private boolean registerJob(JobDetail jobDetail, Trigger trigger) throws SchedulerException
    {
-      if (!jobStoreSupportsPersistence)
+      if (!jobStoreSupportsPersistence) {
          return true;
+      }else{
```
Comment (thomasdelhomenie) [inline exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java]: else is useless here

**Review by thomasdelhomenie — CHANGES_REQUESTED** [review]

**Review by thomasdelhomenie — CHANGES_REQUESTED** [review]
Unit tests are missing.

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java (hunk 1 above)
Comment (aboughzela) [reply exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:674]: No, at start JobSchedulerServiceImpl service , I remove only the list of old  persisted jobs removed from the configuration (persistedJobKeyList contain at start point only the diff between configuration and old persisted list).

Comment (aboughzela) [reply exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:674]: you need to refactoring this code ?

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java (hunk 2 above)
Comment (aboughzela) [reply exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java]: Done

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java (hunk 3 above)
Comment (aboughzela) [reply exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java:729]: Done

File: exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java (hunk 4 above)
Comment (aboughzela) [reply exo.kernel.component.common/src/main/java/org/exoplatform/services/scheduler/impl/JobSchedulerServiceImpl.java]: Done

**Comment by aboughzela** [comment]
> > Unit tests are missing. 
> > The PR  fix a specific scenario , reproduced only with job Stored Persistence before start the standalone container.

**Comment by aboughzela** [comment]
@thomasdelhomenie PR updated
Use scheduler_.rescheduleJob to update only cron job expression or Period Job period time without remove the job

**Review by thomasdelhomenie — APPROVED** [review]

**Pull request closed without merging**